// proto3 문법을 사용함을 선언 (최신 protobuf 버전)
syntax = "proto3";

// 패키지 이름: Rust/Python 코드 생성 시 네임스페이스로 사용됨
package agent;

// ============================================================
// AgentBroker 서비스 정의
// Python(클라이언트) → Rust(서버) 간의 모든 통신 계약을 정의한다.
// 파일 I/O, 명령 실행, 채팅 스트리밍 3가지 카테고리를 제공한다.
// ============================================================
service AgentBroker {

  // [파일 읽기] Python이 파일 경로를 보내면 Rust가 샌드박스 검증 후 내용 반환
  rpc RequestFileRead (FileRequest) returns (FileResponse);

  // [파일 쓰기] Python이 경로+내용을 보내면 Rust가 샌드박스 검증 후 디스크에 기록
  rpc RequestFileWrite (FileWriteRequest) returns (StatusResponse);

  // [명령 실행] Python이 명령어를 보내면 Rust가 화이트리스트 확인 후 실행
  rpc ExecuteCommand (CommandRequest) returns (CommandResponse);

  // [양방향 채팅 스트림] Python ↔ Rust 간 실시간 메시지 스트리밍
  // stream 키워드: 단일 요청/응답이 아닌 연속적인 메시지 흐름을 의미
  rpc StreamChat (stream ChatMessage) returns (stream ChatMessage);
}

// ============================================================
// 메시지 타입 정의
// 각 필드의 숫자(= 1, = 2, ...)는 바이너리 직렬화 시 필드 식별 번호
// ============================================================

// 파일 읽기 요청: 읽고 싶은 파일의 경로만 전달
message FileRequest {
  string path = 1;  // 읽을 파일의 절대/상대 경로
}

// 파일 읽기 응답: 파일 내용 + 허용 여부 + 에러 메시지
message FileResponse {
  bytes content = 1;   // 파일 내용 (바이너리 안전 — 이미지 등도 가능)
  bool allowed = 2;    // 샌드박스 검증 통과 여부 (true=허용, false=거부)
  string error = 3;    // 거부 시 사유 메시지 (allowed=true면 빈 문자열)
}

// 파일 쓰기 요청: 쓸 경로 + 쓸 내용
message FileWriteRequest {
  string path = 1;     // 쓸 파일의 경로
  bytes content = 2;   // 파일에 기록할 바이너리 데이터
}

// 범용 상태 응답: 성공/실패 + 설명 메시지
message StatusResponse {
  bool success = 1;    // 작업 성공 여부
  string message = 2;  // 상세 결과 메시지 (성공 시 확인, 실패 시 에러 내용)
}

// 명령 실행 요청: 실행할 명령어 + 인자 목록
message CommandRequest {
  string command = 1;        // 실행할 명령어 이름 (예: "echo", "ls")
  repeated string args = 2;  // 명령어 인자 배열 (repeated = 0개 이상의 배열)
}

// 명령 실행 응답: 종료 코드 + 표준 출력 + 표준 에러
message CommandResponse {
  int32 exit_code = 1;  // 프로세스 종료 코드 (0=정상, 그 외=에러)
  string stdout = 2;    // 표준 출력 내용
  string stderr = 3;    // 표준 에러 내용
}

// 채팅 메시지: 스트리밍 대화에서 주고받는 단위 메시지
message ChatMessage {
  string role = 1;        // 발화자 역할 ("user", "assistant", "system")
  string content = 2;     // 메시지 본문 텍스트
  string session_id = 3;  // 대화 세션 식별자 (같은 세션의 메시지를 묶는 용도)
}
